### ISSUE TITLE
Fuzion monitors or Fuzion transactional memory

### ISSUE THREAD

**Issue by fridis** [tokiwa-software/fuzion#3997]
We will need  a basic synchronization mechanism that permits flexible but safe concurrent accesses to mutable data larger than single atomic values, e.g., insertions into a binary tree, etc. This should include some notification mechanism to permit, e.g, waiting on an empty queue for some data to be added. Java's monitors are an example, but they have a number of limitations. What we should have

- a means to permit exclusive access to read / write some data
- forbid unprotected access to that data 
- block to wait for data to be updated by another thread
- notify blocked threads that an update occurred

nice to have would be

- clearly specify what data is protected (unlike Java monitors that could protect anything)
- permit several read accesses in parallel 
- show statically that accesses are deadlock free 
- specify on a notify, how many waiting threads should be woken up
- when several threads are blocked waiting to be admitted for an access, wake up these threads in the order of some thread priority
- forbid blocking or I/O operations during an access since these could result in long delays or deadlocks of other threads
- provide deadlines to give up when blocking for an access or when waiting for notification
- allow code to be generic, i.e, written in a way to run with single-threaded `mutate`, multi-threaded `concur.monitor` or multi-threaded `concur.transactional_memory`, etc. 

Idea for a solution 

- add a version of mutate effect that permits to be shared between threads: `concur.monitor` or similar
- this mutate effect should have an `access` function that gives exclusive access to the data to code wrapped in a lambda
- `access` will block in case another thread has exclusive access, will be released once that other thread is done
- the code provided to `access` will be executed with the given `concur.monitor` instance installed, should be usable similar to `mutate`
- additionally, the presence of the `concur.monitor` effect should cause static analysis to flag any blocking operation performed by code run with this effect as an error. 
- a `wait` function similar to Java's `Object.wait` could wait for modifications
- a `notify` function similar to Java's `Object.notify` could wake up 1, n or all `wait`s. 

Alternatively, we could have a version of transactional memory. The main difference being

- an access would not block
- finishing a transaction would atomically update all modified mutable fields (this could be done by a compare-and-swap of a pointer to a structure containing the modifiable data)
- finishing could fail in case of a parallel transaction finished earlier. Then, there could be a choice between retry or giving up

**Comment by fridis** [comment]
This was solved with PR #7067 that added `concur.blocking_mutate` and subsequent enhancements. 

Support for transactional memory via a new version of `mutate` is currently not planned.
